cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java

```diff
@@ -119,4 +146,50 @@ public void validateRepository(NamespaceId namespace, RepositoryConfig repoConfi
       throw new RepositoryConfigValidationException("Internal error: " + e.getMessage(), e);
     }
   }
+
+  public PushAppsResponse pushApps(NamespaceId namespace, List<ApplicationId> appIds, @Nullable String commitMessage)
+    throws Exception {
+    Map<ApplicationId, ApplicationDetail> details = getAndValidateAppDetails(appIds);
+
+    String committer = authenticationContext.getPrincipal().getName();
+    try (RepositoryManager repoManager = createRepositoryManager(namespace)) {
+      SourceControlOperationRunner sourceControlRunner = sourceControlFactory.create(repoManager);
+
+      List<ApplicationDetail> appsToPush = new ArrayList<>(details.values());
+
+      // TODO CDAP-20371 revisit and put correct Author and Committer, for now they are the same
+      CommitMeta commitMeta = new CommitMeta(committer, committer, System.currentTimeMillis(), commitMessage);
+      PushAppsResponse pushResponse = sourceControlRunner.push(appsToPush, commitMeta);
+
+      Map<ApplicationId, SourceControlMeta> pushAppsMeta = pushResponse.getApps().stream()
+        .collect(Collectors.toMap(
+          entry -> namespace.app(entry.getName(), entry.getVersion()),
+          entry -> new SourceControlMeta(entry.getFileHash())
+        ));
+      store.setAppSourceControlMetas(pushAppsMeta);
+      return pushResponse;
+    }
+  }
+
+  private Map<ApplicationId, ApplicationDetail> getAndValidateAppDetails(List<ApplicationId> appIds) throws Exception {
+    // TODO: CDAP-20382 write into disk instead of loading into memory
+    Map<ApplicationId, ApplicationDetail> details = appLifecycleService.getAppDetails(appIds, false);
+
+    // Throw NotFoundException when applications in request is not found.
+    // We don't do partial push and don't send partial result
+    Set<ApplicationId> appsNotFound = Sets.difference(new HashSet<>(appIds), details.keySet());
+    if (!appsNotFound.isEmpty()) {
+      String notFoundMessage = appsNotFound.stream().map(EntityId::toString)
+        .collect(Collectors.joining(", ", "Applications ", " not found."));
+      throw new NotFoundException(notFoundMessage);
```

Comment:
As @tivv suggested, we don't do partial push and return partial result. We throw once app not found detected